{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read/Merge Datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# imports\n",
    "import numpy as np\n",
    "import anndata as ad\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "import copy\n",
    "# import harmonypy as hm\n",
    "# import cellanova as canova\n",
    "from tqdm import tqdm\n",
    "import os\n",
    "# import pyarrow\n",
    "# set matplotlib to show figures inline\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read Satpathy Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 96750 × 31053\n",
       "    var: 'gene_ids'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read data from matrix\n",
    "satpathy_adata = sc.read_10x_mtx(f'../data/Satpathy2022/10X', gex_only=False)\n",
    "\n",
    "# view the adata object\n",
    "satpathy_adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "ename": "PermissionError",
     "evalue": "[Errno 13] Permission denied: '/Genomics/argo/users/ta6403/ParameterEstimation/data/Satpathy2022/metadata.tsv'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mPermissionError\u001b[0m                           Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[5], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# add metadata to adata object\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m satpathy_metadata \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m/Genomics/argo/users/ta6403/ParameterEstimation/data/Satpathy2022/metadata.tsv\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;130;43;01m\\t\u001b[39;49;00m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindex_col\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m c \u001b[38;5;129;01min\u001b[39;00m satpathy_metadata\u001b[38;5;241m.\u001b[39mcolumns:\n\u001b[1;32m      5\u001b[0m     satpathy_adata\u001b[38;5;241m.\u001b[39mobs[c] \u001b[38;5;241m=\u001b[39m satpathy_metadata[c]\n",
      "File \u001b[0;32m/Genomics/pritykinlab/dillon/software/miniconda/envs/envs/scvi/lib/python3.10/site-packages/pandas/util/_decorators.py:311\u001b[0m, in \u001b[0;36mdeprecate_nonkeyword_arguments.<locals>.decorate.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    305\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(args) \u001b[38;5;241m>\u001b[39m num_allow_args:\n\u001b[1;32m    306\u001b[0m     warnings\u001b[38;5;241m.\u001b[39mwarn(\n\u001b[1;32m    307\u001b[0m         msg\u001b[38;5;241m.\u001b[39mformat(arguments\u001b[38;5;241m=\u001b[39marguments),\n\u001b[1;32m    308\u001b[0m         \u001b[38;5;167;01mFutureWarning\u001b[39;00m,\n\u001b[1;32m    309\u001b[0m         stacklevel\u001b[38;5;241m=\u001b[39mstacklevel,\n\u001b[1;32m    310\u001b[0m     )\n\u001b[0;32m--> 311\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/Genomics/pritykinlab/dillon/software/miniconda/envs/envs/scvi/lib/python3.10/site-packages/pandas/io/parsers/readers.py:680\u001b[0m, in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, error_bad_lines, warn_bad_lines, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options)\u001b[0m\n\u001b[1;32m    665\u001b[0m kwds_defaults \u001b[38;5;241m=\u001b[39m _refine_defaults_read(\n\u001b[1;32m    666\u001b[0m     dialect,\n\u001b[1;32m    667\u001b[0m     delimiter,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    676\u001b[0m     defaults\u001b[38;5;241m=\u001b[39m{\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdelimiter\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m,\u001b[39m\u001b[38;5;124m\"\u001b[39m},\n\u001b[1;32m    677\u001b[0m )\n\u001b[1;32m    678\u001b[0m kwds\u001b[38;5;241m.\u001b[39mupdate(kwds_defaults)\n\u001b[0;32m--> 680\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/Genomics/pritykinlab/dillon/software/miniconda/envs/envs/scvi/lib/python3.10/site-packages/pandas/io/parsers/readers.py:575\u001b[0m, in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    572\u001b[0m _validate_names(kwds\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnames\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m))\n\u001b[1;32m    574\u001b[0m \u001b[38;5;66;03m# Create the parser.\u001b[39;00m\n\u001b[0;32m--> 575\u001b[0m parser \u001b[38;5;241m=\u001b[39m \u001b[43mTextFileReader\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    577\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m chunksize \u001b[38;5;129;01mor\u001b[39;00m iterator:\n\u001b[1;32m    578\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n",
      "File \u001b[0;32m/Genomics/pritykinlab/dillon/software/miniconda/envs/envs/scvi/lib/python3.10/site-packages/pandas/io/parsers/readers.py:933\u001b[0m, in \u001b[0;36mTextFileReader.__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m    930\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m kwds[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m    932\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles: IOHandles \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m--> 933\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_engine\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mengine\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/Genomics/pritykinlab/dillon/software/miniconda/envs/envs/scvi/lib/python3.10/site-packages/pandas/io/parsers/readers.py:1217\u001b[0m, in \u001b[0;36mTextFileReader._make_engine\u001b[0;34m(self, f, engine)\u001b[0m\n\u001b[1;32m   1213\u001b[0m     mode \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrb\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1214\u001b[0m \u001b[38;5;66;03m# error: No overload variant of \"get_handle\" matches argument types\u001b[39;00m\n\u001b[1;32m   1215\u001b[0m \u001b[38;5;66;03m# \"Union[str, PathLike[str], ReadCsvBuffer[bytes], ReadCsvBuffer[str]]\"\u001b[39;00m\n\u001b[1;32m   1216\u001b[0m \u001b[38;5;66;03m# , \"str\", \"bool\", \"Any\", \"Any\", \"Any\", \"Any\", \"Any\"\u001b[39;00m\n\u001b[0;32m-> 1217\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;241m=\u001b[39m \u001b[43mget_handle\u001b[49m\u001b[43m(\u001b[49m\u001b[43m  \u001b[49m\u001b[38;5;66;43;03m# type: ignore[call-overload]\u001b[39;49;00m\n\u001b[1;32m   1218\u001b[0m \u001b[43m    \u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1219\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1220\u001b[0m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1221\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcompression\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1222\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmemory_map\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmemory_map\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1223\u001b[0m \u001b[43m    \u001b[49m\u001b[43mis_text\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mis_text\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1224\u001b[0m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding_errors\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstrict\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1225\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstorage_options\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1226\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1227\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m   1228\u001b[0m f \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles\u001b[38;5;241m.\u001b[39mhandle\n",
      "File \u001b[0;32m/Genomics/pritykinlab/dillon/software/miniconda/envs/envs/scvi/lib/python3.10/site-packages/pandas/io/common.py:789\u001b[0m, in \u001b[0;36mget_handle\u001b[0;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[1;32m    784\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(handle, \u001b[38;5;28mstr\u001b[39m):\n\u001b[1;32m    785\u001b[0m     \u001b[38;5;66;03m# Check whether the filename is to be opened in binary mode.\u001b[39;00m\n\u001b[1;32m    786\u001b[0m     \u001b[38;5;66;03m# Binary mode does not support 'encoding' and 'newline'.\u001b[39;00m\n\u001b[1;32m    787\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mencoding \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mmode:\n\u001b[1;32m    788\u001b[0m         \u001b[38;5;66;03m# Encoding\u001b[39;00m\n\u001b[0;32m--> 789\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\n\u001b[1;32m    790\u001b[0m \u001b[43m            \u001b[49m\u001b[43mhandle\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    791\u001b[0m \u001b[43m            \u001b[49m\u001b[43mioargs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    792\u001b[0m \u001b[43m            \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mioargs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    793\u001b[0m \u001b[43m            \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    794\u001b[0m \u001b[43m            \u001b[49m\u001b[43mnewline\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    795\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    796\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    797\u001b[0m         \u001b[38;5;66;03m# Binary mode\u001b[39;00m\n\u001b[1;32m    798\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mopen\u001b[39m(handle, ioargs\u001b[38;5;241m.\u001b[39mmode)\n",
      "\u001b[0;31mPermissionError\u001b[0m: [Errno 13] Permission denied: '/Genomics/argo/users/ta6403/ParameterEstimation/data/Satpathy2022/metadata.tsv'"
     ]
    }
   ],
   "source": [
    "# add metadata to adata object\n",
    "satpathy_metadata = pd.read_csv(\"/Genomics/argo/users/ta6403/ParameterEstimation/data/Satpathy2022/metadata.tsv\", sep=\"\\t\", index_col=0)\n",
    "\n",
    "for c in satpathy_metadata.columns:\n",
    "    satpathy_adata.obs[c] = satpathy_metadata[c]\n",
    "\n",
    "# experiment defined in Satpathy's github\n",
    "satpathy_adata.obs['experiment'] = 'exp1'\n",
    "exp2 = satpathy_adata.obs['orig.ident'].isin([\"Spleen_Chronic_D21_PD1Pos\"\n",
    "                                 , \"Spleen_Chronic_D21_CX3CR1Pos\"\n",
    "                                 , \"Spleen_Chronic_D21_CX3CR1Neg_Slamf6Pos\"\n",
    "                                 , \"Spleen_Chronic_D21_CX3CR1Neg\"\n",
    "                                 , \"Spleen_Chronic_D21_TetPos_rep1\"\n",
    "                                 , \"Spleen_Chronic_D21_TetNeg_rep1\"])\n",
    "satpathy_adata.obs.loc[exp2.to_numpy(), 'experiment'] = 'exp2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# view metadata\n",
    "satpathy_adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# subset only spleen chronic day 21 cells\n",
    "satpathy_adata = satpathy_adata[satpathy_adata.obs.lcmv=='Chronic_D21']\n",
    "# satpathy_adata = satpathy_adata[satpathy_adata.obs.tissue=='Spleen']\n",
    "\n",
    "# filter only cells with both TRA and TRB\n",
    "# satpathy_adata = satpathy_adata[~satpathy_adata.obs['tcr_cdr3s_aa'].isna()]\n",
    "# satpathy_adata = satpathy_adata[satpathy_adata.obs['tcr_cdr3s_aa'].str.contains('TRA:')]\n",
    "# satpathy_adata = satpathy_adata[satpathy_adata.obs['tcr_cdr3s_aa'].str.contains('TRB:')]\n",
    "\n",
    "satpathy_adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "satpathy_adata.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read Cui Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Genomics/argo/users/ta6403/.conda/envs/ta6403/lib/python3.11/site-packages/anndata/_core/anndata.py:1763: FutureWarning: The AnnData.concatenate method is deprecated in favour of the anndata.concat function. Please use anndata.concat instead.\n",
      "\n",
      "See the tutorial for concat at: https://anndata.readthedocs.io/en/latest/concatenation.html\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 9094 × 31053\n",
       "    obs: 'batch', 'cell_id'\n",
       "    var: 'gene_ids', 'feature_types'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read data from Cui 2023\n",
    "cui_adatas= [sc.read_10x_mtx(f'/Genomics/argo/users/ta6403/TExhaustion/Reanalysis/Cui2023/M{i+1}', gex_only=False) for i in range(3)]\n",
    "# correct for file processed in v2\n",
    "cui_adatas[-1].var['feature_types'] = ['Gene Expression'] * len(cui_adatas[-1].var['gene_ids'])\n",
    "\n",
    "\n",
    "# concatenate all data and save as adata\n",
    "cui_adata = cui_adatas[0].concatenate(cui_adatas[1:])\n",
    "\n",
    "cui_adata.obs['cell_id'] = range(len(cui_adata.obs))\n",
    "\n",
    "# view the adata object\n",
    "cui_adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>batch</th>\n",
       "      <th>cell_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGCTCCTTC-1-0</th>\n",
       "      <td>M1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGATGCGT-1-0</th>\n",
       "      <td>M1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGTAAACT-1-0</th>\n",
       "      <td>M1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGGTGGTTTCA-1-0</th>\n",
       "      <td>M1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGGTTATCACG-1-0</th>\n",
       "      <td>M1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTTCTCAAACG-1-2</th>\n",
       "      <td>M3</td>\n",
       "      <td>9089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAAGATGTCGG-1-2</th>\n",
       "      <td>M3</td>\n",
       "      <td>9090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCACACGGACAA-1-2</th>\n",
       "      <td>M3</td>\n",
       "      <td>9091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCACATGGGACA-1-2</th>\n",
       "      <td>M3</td>\n",
       "      <td>9092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTAGTACCT-1-2</th>\n",
       "      <td>M3</td>\n",
       "      <td>9093</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9094 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     batch  cell_id\n",
       "AAACCTGAGCTCCTTC-1-0    M1        0\n",
       "AAACCTGAGGATGCGT-1-0    M1        1\n",
       "AAACCTGAGGTAAACT-1-0    M1        2\n",
       "AAACCTGGTGGTTTCA-1-0    M1        3\n",
       "AAACCTGGTTATCACG-1-0    M1        4\n",
       "...                    ...      ...\n",
       "TTTGGTTTCTCAAACG-1-2    M3     9089\n",
       "TTTGTCAAGATGTCGG-1-2    M3     9090\n",
       "TTTGTCACACGGACAA-1-2    M3     9091\n",
       "TTTGTCACATGGGACA-1-2    M3     9092\n",
       "TTTGTCAGTAGTACCT-1-2    M3     9093\n",
       "\n",
       "[9094 rows x 2 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create dictionary mapping batch numbers to names\n",
    "batch_name_dict = {'0':'M1', '1':'M2', '2':'M3'}\n",
    "\n",
    "# use dictionary to replace batch numbers with names\n",
    "cui_adata.obs = cui_adata.obs.replace({'batch': batch_name_dict})\n",
    "\n",
    "# view updated obs\n",
    "cui_adata.obs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  2%|▏         | 155/8273 [00:00<00:10, 770.93it/s]100%|██████████| 8273/8273 [00:10<00:00, 768.61it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tcr_cdr3s_aa</th>\n",
       "      <th>tcr_cdr3s_nt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGATGCGT-1-0</th>\n",
       "      <td>TRA:CAAREGGSALGRLHF;TRA:CAASPDNNAPRF;TRB:CASSD...</td>\n",
       "      <td>TRA:TGTGCAGCAAGGGAGGGAGGTTCAGCCTTAGGGAGGCTGCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGTAAACT-1-0</th>\n",
       "      <td>TRA:CAMREGGGTGSKLSF;TRB:CASSIWDGQNTLYF</td>\n",
       "      <td>TRA:TGTGCTATGAGAGAGGGCGGAGGCACTGGGTCTAAGCTGTCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGGTGGTTTCA-1-0</th>\n",
       "      <td>TRA:CAVSPDYSNNRLTL;TRB:CASRANSYNSPLYF</td>\n",
       "      <td>TRA:TGTGCTGTGAGCCCGGACTACAGCAACAACAGACTTACTTTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGGTTATCACG-1-0</th>\n",
       "      <td>TRA:CAMREGTEGADRLTF;TRB:CASSLIQGRYAEQFF</td>\n",
       "      <td>TRA:TGTGCTATGAGAGAGGGAACAGAAGGTGCAGATAGACTCACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGTCACATAGC-1-0</th>\n",
       "      <td>TRA:CAAYTSSSFSKLVF;TRB:CASSTWGNTGQLYF</td>\n",
       "      <td>TRA:TGTGCTGCATACACATCCTCCTCCTTCAGCAAGCTGGTGTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGCGCTCATGTCTT-1-2</th>\n",
       "      <td>TRA:CAVSDSGTYQRF;TRB:CASSPGQYNSPLYF</td>\n",
       "      <td>TRA:TGTGCTGTGAGCGATTCTGGGACTTACCAGAGGTTT;TRB:T...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTGTCAGCTAT-1-2</th>\n",
       "      <td>TRA:CAIEDSNYQLIW;TRB:CSADHGNYAEQFF;TRB:CASSPGQ...</td>\n",
       "      <td>TRA:TGTGCTATAGAGGATAGCAACTATCAGTTGATCTGG;TRB:T...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTTCTCAAACG-1-2</th>\n",
       "      <td>TRA:CAVSDSGTYQRF;TRB:CASSPGQYNSPLYF</td>\n",
       "      <td>TRA:TGTGCTGTGAGCGATTCTGGGACTTACCAGAGGTTT;TRB:T...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAAGATGTCGG-1-2</th>\n",
       "      <td>TRA:CAASGGSALGRLHF;TRB:CASSRKRGGQNTLYF</td>\n",
       "      <td>TRA:TGTGCAGCAAGTGGCGGTTCAGCCTTAGGGAGGCTGCATTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTAGTACCT-1-2</th>\n",
       "      <td>TRA:CAVVGANTGKLTF;TRB:CASSDHANTEVFF</td>\n",
       "      <td>TRA:TGTGCAGTGGTTGGAGCTAACACTGGAAAGCTCACGTTT;TR...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5260 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                           tcr_cdr3s_aa  \\\n",
       "AAACCTGAGGATGCGT-1-0  TRA:CAAREGGSALGRLHF;TRA:CAASPDNNAPRF;TRB:CASSD...   \n",
       "AAACCTGAGGTAAACT-1-0             TRA:CAMREGGGTGSKLSF;TRB:CASSIWDGQNTLYF   \n",
       "AAACCTGGTGGTTTCA-1-0              TRA:CAVSPDYSNNRLTL;TRB:CASRANSYNSPLYF   \n",
       "AAACCTGGTTATCACG-1-0            TRA:CAMREGTEGADRLTF;TRB:CASSLIQGRYAEQFF   \n",
       "AAACCTGTCACATAGC-1-0              TRA:CAAYTSSSFSKLVF;TRB:CASSTWGNTGQLYF   \n",
       "...                                                                 ...   \n",
       "TTTGCGCTCATGTCTT-1-2                TRA:CAVSDSGTYQRF;TRB:CASSPGQYNSPLYF   \n",
       "TTTGGTTGTCAGCTAT-1-2  TRA:CAIEDSNYQLIW;TRB:CSADHGNYAEQFF;TRB:CASSPGQ...   \n",
       "TTTGGTTTCTCAAACG-1-2                TRA:CAVSDSGTYQRF;TRB:CASSPGQYNSPLYF   \n",
       "TTTGTCAAGATGTCGG-1-2             TRA:CAASGGSALGRLHF;TRB:CASSRKRGGQNTLYF   \n",
       "TTTGTCAGTAGTACCT-1-2                TRA:CAVVGANTGKLTF;TRB:CASSDHANTEVFF   \n",
       "\n",
       "                                                           tcr_cdr3s_nt  \n",
       "AAACCTGAGGATGCGT-1-0  TRA:TGTGCAGCAAGGGAGGGAGGTTCAGCCTTAGGGAGGCTGCAT...  \n",
       "AAACCTGAGGTAAACT-1-0  TRA:TGTGCTATGAGAGAGGGCGGAGGCACTGGGTCTAAGCTGTCA...  \n",
       "AAACCTGGTGGTTTCA-1-0  TRA:TGTGCTGTGAGCCCGGACTACAGCAACAACAGACTTACTTTG...  \n",
       "AAACCTGGTTATCACG-1-0  TRA:TGTGCTATGAGAGAGGGAACAGAAGGTGCAGATAGACTCACC...  \n",
       "AAACCTGTCACATAGC-1-0  TRA:TGTGCTGCATACACATCCTCCTCCTTCAGCAAGCTGGTGTTT...  \n",
       "...                                                                 ...  \n",
       "TTTGCGCTCATGTCTT-1-2  TRA:TGTGCTGTGAGCGATTCTGGGACTTACCAGAGGTTT;TRB:T...  \n",
       "TTTGGTTGTCAGCTAT-1-2  TRA:TGTGCTATAGAGGATAGCAACTATCAGTTGATCTGG;TRB:T...  \n",
       "TTTGGTTTCTCAAACG-1-2  TRA:TGTGCTGTGAGCGATTCTGGGACTTACCAGAGGTTT;TRB:T...  \n",
       "TTTGTCAAGATGTCGG-1-2  TRA:TGTGCAGCAAGTGGCGGTTCAGCCTTAGGGAGGCTGCATTTT...  \n",
       "TTTGTCAGTAGTACCT-1-2  TRA:TGTGCAGTGGTTGGAGCTAACACTGGAAAGCTCACGTTT;TR...  \n",
       "\n",
       "[5260 rows x 2 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read Cui 2023 metadata\n",
    "cui_metadata = pd.read_csv('../data/Cui2023/cui_metadata.tsv', sep='\\t')\n",
    "\n",
    "# filter to only those with TCRs\n",
    "cui_metadata = cui_metadata[~cui_metadata['cdr3'].isna()]\n",
    "cui_metadata = cui_metadata[cui_metadata['chain']!='Multi']\n",
    "cui_metadata = cui_metadata[cui_metadata['productive']==True]\n",
    "\n",
    "barcodes = []\n",
    "tcrs_aa = []\n",
    "tcrs_nt = []\n",
    "mice = []\n",
    "\n",
    "# TRA and TRB are separate rows of the dataset, so we combine them into one dataframe with TRA/TRB as one row\n",
    "for barcode in tqdm(cui_metadata['barcode'].unique()):\n",
    "    cui_subset = cui_metadata[cui_metadata.barcode==barcode]\n",
    "    vcs = cui_subset.chain.value_counts()\n",
    "    if len(vcs)<2 or len(vcs)>3:\n",
    "        continue\n",
    "    cui_subset = cui_subset.sort_values('chain')\n",
    "    mouse_int = int(cui_subset.mouse.iloc[0][1])-1\n",
    "    barcode = barcode+f'-{mouse_int}'\n",
    "    if vcs['TRA']==1 and vcs['TRB']==1:\n",
    "        tcrs_aa.append(f'TRA:{cui_subset.cdr3.iloc[0]};TRB:{cui_subset.cdr3.iloc[1]}')\n",
    "        tcrs_nt.append(f'TRA:{cui_subset.cdr3_nt.iloc[0]};TRB:{cui_subset.cdr3_nt.iloc[1]}')\n",
    "        barcodes.append(barcode) # barcode the same index as the dataframe\n",
    "    if vcs['TRA']==2 and vcs['TRB']==1:\n",
    "        tcrs_aa.append(f'TRA:{cui_subset.cdr3.iloc[0]};TRA:{cui_subset.cdr3.iloc[1]};TRB:{cui_subset.cdr3.iloc[2]}')\n",
    "        tcrs_nt.append(f'TRA:{cui_subset.cdr3_nt.iloc[0]};TRA:{cui_subset.cdr3_nt.iloc[1]};TRB:{cui_subset.cdr3_nt.iloc[2]}')\n",
    "        barcodes.append(barcode) # barcode the same index as the dataframe\n",
    "    if vcs['TRB']==2 and vcs['TRA']==1:\n",
    "        tcrs_aa.append(f'TRA:{cui_subset.cdr3.iloc[0]};TRB:{cui_subset.cdr3.iloc[1]};TRB:{cui_subset.cdr3.iloc[2]}')\n",
    "        tcrs_nt.append(f'TRA:{cui_subset.cdr3_nt.iloc[0]};TRB:{cui_subset.cdr3_nt.iloc[1]};TRB:{cui_subset.cdr3_nt.iloc[2]}')\n",
    "        barcodes.append(barcode) # barcode the same index as the dataframe\n",
    "\n",
    "\n",
    "\n",
    "cui_metadata = pd.DataFrame({'tcr_cdr3s_aa': tcrs_aa, 'tcr_cdr3s_nt': tcrs_nt}, index=barcodes)\n",
    "cui_metadata\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>batch</th>\n",
       "      <th>cell_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGCTCCTTC-1-0</th>\n",
       "      <td>M1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGATGCGT-1-0</th>\n",
       "      <td>M1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGTAAACT-1-0</th>\n",
       "      <td>M1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGGTGGTTTCA-1-0</th>\n",
       "      <td>M1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGGTTATCACG-1-0</th>\n",
       "      <td>M1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTTCTCAAACG-1-2</th>\n",
       "      <td>M3</td>\n",
       "      <td>9089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAAGATGTCGG-1-2</th>\n",
       "      <td>M3</td>\n",
       "      <td>9090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCACACGGACAA-1-2</th>\n",
       "      <td>M3</td>\n",
       "      <td>9091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCACATGGGACA-1-2</th>\n",
       "      <td>M3</td>\n",
       "      <td>9092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTAGTACCT-1-2</th>\n",
       "      <td>M3</td>\n",
       "      <td>9093</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9094 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     batch  cell_id\n",
       "AAACCTGAGCTCCTTC-1-0    M1        0\n",
       "AAACCTGAGGATGCGT-1-0    M1        1\n",
       "AAACCTGAGGTAAACT-1-0    M1        2\n",
       "AAACCTGGTGGTTTCA-1-0    M1        3\n",
       "AAACCTGGTTATCACG-1-0    M1        4\n",
       "...                    ...      ...\n",
       "TTTGGTTTCTCAAACG-1-2    M3     9089\n",
       "TTTGTCAAGATGTCGG-1-2    M3     9090\n",
       "TTTGTCACACGGACAA-1-2    M3     9091\n",
       "TTTGTCACATGGGACA-1-2    M3     9092\n",
       "TTTGTCAGTAGTACCT-1-2    M3     9093\n",
       "\n",
       "[9094 rows x 2 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cui_adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add TCRs back to adata\n",
    "cui_merged = pd.merge(cui_metadata, cui_adata.obs, how='right', left_index=True, right_index=True)\n",
    "cui_adata = cui_adata[cui_merged.cell_id]\n",
    "cui_adata.obs = cui_merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tcr_cdr3s_aa</th>\n",
       "      <th>tcr_cdr3s_nt</th>\n",
       "      <th>batch</th>\n",
       "      <th>cell_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGCTCCTTC-1-0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>M1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGATGCGT-1-0</th>\n",
       "      <td>TRA:CAAREGGSALGRLHF;TRA:CAASPDNNAPRF;TRB:CASSD...</td>\n",
       "      <td>TRA:TGTGCAGCAAGGGAGGGAGGTTCAGCCTTAGGGAGGCTGCAT...</td>\n",
       "      <td>M1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGTAAACT-1-0</th>\n",
       "      <td>TRA:CAMREGGGTGSKLSF;TRB:CASSIWDGQNTLYF</td>\n",
       "      <td>TRA:TGTGCTATGAGAGAGGGCGGAGGCACTGGGTCTAAGCTGTCA...</td>\n",
       "      <td>M1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGGTGGTTTCA-1-0</th>\n",
       "      <td>TRA:CAVSPDYSNNRLTL;TRB:CASRANSYNSPLYF</td>\n",
       "      <td>TRA:TGTGCTGTGAGCCCGGACTACAGCAACAACAGACTTACTTTG...</td>\n",
       "      <td>M1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGGTTATCACG-1-0</th>\n",
       "      <td>TRA:CAMREGTEGADRLTF;TRB:CASSLIQGRYAEQFF</td>\n",
       "      <td>TRA:TGTGCTATGAGAGAGGGAACAGAAGGTGCAGATAGACTCACC...</td>\n",
       "      <td>M1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTTCTCAAACG-1-2</th>\n",
       "      <td>TRA:CAVSDSGTYQRF;TRB:CASSPGQYNSPLYF</td>\n",
       "      <td>TRA:TGTGCTGTGAGCGATTCTGGGACTTACCAGAGGTTT;TRB:T...</td>\n",
       "      <td>M3</td>\n",
       "      <td>9089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAAGATGTCGG-1-2</th>\n",
       "      <td>TRA:CAASGGSALGRLHF;TRB:CASSRKRGGQNTLYF</td>\n",
       "      <td>TRA:TGTGCAGCAAGTGGCGGTTCAGCCTTAGGGAGGCTGCATTTT...</td>\n",
       "      <td>M3</td>\n",
       "      <td>9090</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCACACGGACAA-1-2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>M3</td>\n",
       "      <td>9091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCACATGGGACA-1-2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>M3</td>\n",
       "      <td>9092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTAGTACCT-1-2</th>\n",
       "      <td>TRA:CAVVGANTGKLTF;TRB:CASSDHANTEVFF</td>\n",
       "      <td>TRA:TGTGCAGTGGTTGGAGCTAACACTGGAAAGCTCACGTTT;TR...</td>\n",
       "      <td>M3</td>\n",
       "      <td>9093</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9094 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                           tcr_cdr3s_aa  \\\n",
       "AAACCTGAGCTCCTTC-1-0                                                NaN   \n",
       "AAACCTGAGGATGCGT-1-0  TRA:CAAREGGSALGRLHF;TRA:CAASPDNNAPRF;TRB:CASSD...   \n",
       "AAACCTGAGGTAAACT-1-0             TRA:CAMREGGGTGSKLSF;TRB:CASSIWDGQNTLYF   \n",
       "AAACCTGGTGGTTTCA-1-0              TRA:CAVSPDYSNNRLTL;TRB:CASRANSYNSPLYF   \n",
       "AAACCTGGTTATCACG-1-0            TRA:CAMREGTEGADRLTF;TRB:CASSLIQGRYAEQFF   \n",
       "...                                                                 ...   \n",
       "TTTGGTTTCTCAAACG-1-2                TRA:CAVSDSGTYQRF;TRB:CASSPGQYNSPLYF   \n",
       "TTTGTCAAGATGTCGG-1-2             TRA:CAASGGSALGRLHF;TRB:CASSRKRGGQNTLYF   \n",
       "TTTGTCACACGGACAA-1-2                                                NaN   \n",
       "TTTGTCACATGGGACA-1-2                                                NaN   \n",
       "TTTGTCAGTAGTACCT-1-2                TRA:CAVVGANTGKLTF;TRB:CASSDHANTEVFF   \n",
       "\n",
       "                                                           tcr_cdr3s_nt batch  \\\n",
       "AAACCTGAGCTCCTTC-1-0                                                NaN    M1   \n",
       "AAACCTGAGGATGCGT-1-0  TRA:TGTGCAGCAAGGGAGGGAGGTTCAGCCTTAGGGAGGCTGCAT...    M1   \n",
       "AAACCTGAGGTAAACT-1-0  TRA:TGTGCTATGAGAGAGGGCGGAGGCACTGGGTCTAAGCTGTCA...    M1   \n",
       "AAACCTGGTGGTTTCA-1-0  TRA:TGTGCTGTGAGCCCGGACTACAGCAACAACAGACTTACTTTG...    M1   \n",
       "AAACCTGGTTATCACG-1-0  TRA:TGTGCTATGAGAGAGGGAACAGAAGGTGCAGATAGACTCACC...    M1   \n",
       "...                                                                 ...   ...   \n",
       "TTTGGTTTCTCAAACG-1-2  TRA:TGTGCTGTGAGCGATTCTGGGACTTACCAGAGGTTT;TRB:T...    M3   \n",
       "TTTGTCAAGATGTCGG-1-2  TRA:TGTGCAGCAAGTGGCGGTTCAGCCTTAGGGAGGCTGCATTTT...    M3   \n",
       "TTTGTCACACGGACAA-1-2                                                NaN    M3   \n",
       "TTTGTCACATGGGACA-1-2                                                NaN    M3   \n",
       "TTTGTCAGTAGTACCT-1-2  TRA:TGTGCAGTGGTTGGAGCTAACACTGGAAAGCTCACGTTT;TR...    M3   \n",
       "\n",
       "                      cell_id  \n",
       "AAACCTGAGCTCCTTC-1-0        0  \n",
       "AAACCTGAGGATGCGT-1-0        1  \n",
       "AAACCTGAGGTAAACT-1-0        2  \n",
       "AAACCTGGTGGTTTCA-1-0        3  \n",
       "AAACCTGGTTATCACG-1-0        4  \n",
       "...                       ...  \n",
       "TTTGGTTTCTCAAACG-1-2     9089  \n",
       "TTTGTCAAGATGTCGG-1-2     9090  \n",
       "TTTGTCACACGGACAA-1-2     9091  \n",
       "TTTGTCACATGGGACA-1-2     9092  \n",
       "TTTGTCAGTAGTACCT-1-2     9093  \n",
       "\n",
       "[9094 rows x 4 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cui_adata.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read Yermanos Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 4503 × 32285\n",
       "    var: 'gene_ids', 'feature_types'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read data from matrix\n",
    "yermanos_adata = sc.read_10x_mtx(f'../data/Yermanos2022/10X', gex_only=False)\n",
    "\n",
    "# view the adata object\n",
    "yermanos_adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>cell_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAGGACA-1</th>\n",
       "      <td>sample3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCACAGCGTC-1</th>\n",
       "      <td>sample3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCACCGAATT-1</th>\n",
       "      <td>sample3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGGTAGCGCTC-1</th>\n",
       "      <td>sample3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGTCCAGGGCT-1</th>\n",
       "      <td>sample3</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTTCATCACCC-2</th>\n",
       "      <td>sample4</td>\n",
       "      <td>4498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAAGTACATGA-2</th>\n",
       "      <td>sample4</td>\n",
       "      <td>4499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCACACCATCCT-2</th>\n",
       "      <td>sample4</td>\n",
       "      <td>4500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTGCAGACA-2</th>\n",
       "      <td>sample4</td>\n",
       "      <td>4501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTGTTGGGA-2</th>\n",
       "      <td>sample4</td>\n",
       "      <td>4502</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4503 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     sample  cell_id\n",
       "AAACCTGAGAAGGACA-1  sample3        0\n",
       "AAACCTGCACAGCGTC-1  sample3        1\n",
       "AAACCTGCACCGAATT-1  sample3        2\n",
       "AAACCTGGTAGCGCTC-1  sample3        3\n",
       "AAACCTGTCCAGGGCT-1  sample3        4\n",
       "...                     ...      ...\n",
       "TTTGGTTTCATCACCC-2  sample4     4498\n",
       "TTTGTCAAGTACATGA-2  sample4     4499\n",
       "TTTGTCACACCATCCT-2  sample4     4500\n",
       "TTTGTCAGTGCAGACA-2  sample4     4501\n",
       "TTTGTCAGTGTTGGGA-2  sample4     4502\n",
       "\n",
       "[4503 rows x 2 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "barcodes = pd.read_csv('../data/Yermanos2022/10X/barcodes.tsv.gz', header=None)\n",
    "\n",
    "yermanos_adata.obs['barcode'] = np.array(barcodes[0])\n",
    "\n",
    "yermanos_adata.obs.set_index('barcode')\n",
    "yermanos_adata.obs['sample'] = 'sample'+(yermanos_adata.obs.barcode.str[-1].astype(int)+2).astype(str)\n",
    "\n",
    "yermanos_adata.obs['cell_id'] = np.arange(len(yermanos_adata))\n",
    "\n",
    "del yermanos_adata.obs['barcode']\n",
    "yermanos_adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  5%|▌         | 136/2640 [00:00<00:01, 1355.97it/s]100%|██████████| 2640/2640 [00:01<00:00, 1377.99it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tcr_cdr3s_aa</th>\n",
       "      <th>tcr_cdr3s_nt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAGGACA-1</th>\n",
       "      <td>TRA:CAVSAGSNYNVLYF;TRB:CASSFWGDSSYEQYF;TRB:CAS...</td>\n",
       "      <td>TRA:TGCGCAGTCAGTGCGGGGTCTAATTACAACGTGCTTTACTTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGTCGTGACAT-1</th>\n",
       "      <td>TRA:CDSGTYQRF;TRB:CASRDRGRNSPLYF;TRB:CASSFWGDS...</td>\n",
       "      <td>TRA:TGTGATTCTGGGACTTACCAGAGGTTT;TRB:TGTGCCAGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGGGTCCACGCAG-1</th>\n",
       "      <td>TRA:CAVRDNYAQGLTF;TRB:CASSSRTTNSDYTF</td>\n",
       "      <td>TRA:TGTGCTGTGAGGGATAACTATGCCCAGGGATTAACCTTC;TR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAGATGAGAGTAAGG-1</th>\n",
       "      <td>TRA:CAVRDNYAQGLTF;TRB:CASSSRTTNSDYTF</td>\n",
       "      <td>TRA:TGTGCTGTGAGGGATAACTATGCCCAGGGATTAACCTTC;TR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAGATGAGGAATTAC-1</th>\n",
       "      <td>TRA:CAASSDTNAYKVIF;TRB:CASKTGGPYEQYF</td>\n",
       "      <td>TRA:TGTGCAGCTAGTTCTGACACAAATGCTTACAAAGTCATCTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTAGAGAACAG-2</th>\n",
       "      <td>TRA:CAMREITGNTGKLIF;TRB:CASSLLYEQYF</td>\n",
       "      <td>TRA:TGTGCTATGAGAGAGATAACAGGCAATACCGGAAAACTCATC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTAGAGTACCG-2</th>\n",
       "      <td>TRA:CAANYGNEKITF;TRB:CASSFVTNQDTQYF</td>\n",
       "      <td>TRA:TGTGCTGCAAACTATGGAAATGAGAAAATAACTTTT;TRB:T...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTGTCGACTGC-2</th>\n",
       "      <td>TRA:CALGFMGYKLTF;TRB:CASSIWQGTNERLFF</td>\n",
       "      <td>TRA:TGTGCTCTGGGTTTCATGGGCTACAAACTTACCTTC;TRB:T...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAAGTACATGA-2</th>\n",
       "      <td>TRA:CAAIYPGYQNFYF;TRB:CASSDNGNTEVFF</td>\n",
       "      <td>TRA:TGTGCAGCAATCTATCCGGGTTACCAGAACTTCTATTTT;TR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCACACCATCCT-2</th>\n",
       "      <td>TRA:CAAIYPGYQNFYF;TRB:CASSDHTNTEVFF</td>\n",
       "      <td>TRA:TGTGCAGCAATCTATCCGGGTTACCAGAACTTCTATTTT;TR...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1663 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                         tcr_cdr3s_aa  \\\n",
       "AAACCTGAGAAGGACA-1  TRA:CAVSAGSNYNVLYF;TRB:CASSFWGDSSYEQYF;TRB:CAS...   \n",
       "AAACCTGTCGTGACAT-1  TRA:CDSGTYQRF;TRB:CASRDRGRNSPLYF;TRB:CASSFWGDS...   \n",
       "AAACGGGTCCACGCAG-1               TRA:CAVRDNYAQGLTF;TRB:CASSSRTTNSDYTF   \n",
       "AAAGATGAGAGTAAGG-1               TRA:CAVRDNYAQGLTF;TRB:CASSSRTTNSDYTF   \n",
       "AAAGATGAGGAATTAC-1               TRA:CAASSDTNAYKVIF;TRB:CASKTGGPYEQYF   \n",
       "...                                                               ...   \n",
       "TTTGGTTAGAGAACAG-2                TRA:CAMREITGNTGKLIF;TRB:CASSLLYEQYF   \n",
       "TTTGGTTAGAGTACCG-2                TRA:CAANYGNEKITF;TRB:CASSFVTNQDTQYF   \n",
       "TTTGGTTGTCGACTGC-2               TRA:CALGFMGYKLTF;TRB:CASSIWQGTNERLFF   \n",
       "TTTGTCAAGTACATGA-2                TRA:CAAIYPGYQNFYF;TRB:CASSDNGNTEVFF   \n",
       "TTTGTCACACCATCCT-2                TRA:CAAIYPGYQNFYF;TRB:CASSDHTNTEVFF   \n",
       "\n",
       "                                                         tcr_cdr3s_nt  \n",
       "AAACCTGAGAAGGACA-1  TRA:TGCGCAGTCAGTGCGGGGTCTAATTACAACGTGCTTTACTTC...  \n",
       "AAACCTGTCGTGACAT-1  TRA:TGTGATTCTGGGACTTACCAGAGGTTT;TRB:TGTGCCAGCA...  \n",
       "AAACGGGTCCACGCAG-1  TRA:TGTGCTGTGAGGGATAACTATGCCCAGGGATTAACCTTC;TR...  \n",
       "AAAGATGAGAGTAAGG-1  TRA:TGTGCTGTGAGGGATAACTATGCCCAGGGATTAACCTTC;TR...  \n",
       "AAAGATGAGGAATTAC-1  TRA:TGTGCAGCTAGTTCTGACACAAATGCTTACAAAGTCATCTTT...  \n",
       "...                                                               ...  \n",
       "TTTGGTTAGAGAACAG-2  TRA:TGTGCTATGAGAGAGATAACAGGCAATACCGGAAAACTCATC...  \n",
       "TTTGGTTAGAGTACCG-2  TRA:TGTGCTGCAAACTATGGAAATGAGAAAATAACTTTT;TRB:T...  \n",
       "TTTGGTTGTCGACTGC-2  TRA:TGTGCTCTGGGTTTCATGGGCTACAAACTTACCTTC;TRB:T...  \n",
       "TTTGTCAAGTACATGA-2  TRA:TGTGCAGCAATCTATCCGGGTTACCAGAACTTCTATTTT;TR...  \n",
       "TTTGTCACACCATCCT-2  TRA:TGTGCAGCAATCTATCCGGGTTACCAGAACTTCTATTTT;TR...  \n",
       "\n",
       "[1663 rows x 2 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read Cui 2023 metadata\n",
    "yermanos_metadata = pd.read_csv('../data/Yermanos2022/10X/metadata.csv')\n",
    "\n",
    "# filter to only those with TCRs\n",
    "yermanos_metadata = yermanos_metadata[~yermanos_metadata['cdr3'].isna()]\n",
    "yermanos_metadata = yermanos_metadata[yermanos_metadata['chain']!='Multi']\n",
    "yermanos_metadata = yermanos_metadata[yermanos_metadata['productive']==True]\n",
    "\n",
    "barcodes = []\n",
    "tcrs_aa = []\n",
    "tcrs_nt = []\n",
    "\n",
    "# TRA and TRB are separate rows of the dataset, so we combine them into one dataframe with TRA/TRB as one row\n",
    "for barcode in tqdm(yermanos_metadata['barcode'].unique()):\n",
    "    yermanos_subset = yermanos_metadata[yermanos_metadata.barcode==barcode]\n",
    "    vcs = yermanos_subset.chain.value_counts()\n",
    "    if len(vcs)<2 or len(vcs)>3:\n",
    "        continue\n",
    "    yermanos_subset = yermanos_subset.sort_values('chain')\n",
    "    if vcs['TRA']==1 and vcs['TRB']==1:\n",
    "        tcrs_aa.append(f'TRA:{yermanos_subset.cdr3.iloc[0]};TRB:{yermanos_subset.cdr3.iloc[1]}')\n",
    "        tcrs_nt.append(f'TRA:{yermanos_subset.cdr3_nt.iloc[0]};TRB:{yermanos_subset.cdr3_nt.iloc[1]}')\n",
    "        barcodes.append(barcode) # barcode the same index as the dataframe\n",
    "    if vcs['TRA']==2 and vcs['TRB']==1:\n",
    "        tcrs_aa.append(f'TRA:{yermanos_subset.cdr3.iloc[0]};TRA:{yermanos_subset.cdr3.iloc[1]};TRB:{yermanos_subset.cdr3.iloc[2]}')\n",
    "        tcrs_nt.append(f'TRA:{yermanos_subset.cdr3_nt.iloc[0]};TRA:{yermanos_subset.cdr3_nt.iloc[1]};TRB:{yermanos_subset.cdr3_nt.iloc[2]}')\n",
    "        barcodes.append(barcode) # barcode the same index as the dataframe\n",
    "    if vcs['TRB']==2 and vcs['TRA']==1:\n",
    "        tcrs_aa.append(f'TRA:{yermanos_subset.cdr3.iloc[0]};TRB:{yermanos_subset.cdr3.iloc[1]};TRB:{yermanos_subset.cdr3.iloc[2]}')\n",
    "        tcrs_nt.append(f'TRA:{yermanos_subset.cdr3_nt.iloc[0]};TRB:{yermanos_subset.cdr3_nt.iloc[1]};TRB:{yermanos_subset.cdr3_nt.iloc[2]}')\n",
    "        barcodes.append(barcode) # barcode the same index as the dataframe\n",
    "\n",
    "yermanos_metadata = pd.DataFrame({'tcr_cdr3s_aa': tcrs_aa, 'tcr_cdr3s_nt': tcrs_nt}, index=barcodes)\n",
    "yermanos_metadata\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add TCRs back to adata\n",
    "yermanos_merged = pd.merge(yermanos_metadata, yermanos_adata.obs, how='right', left_index=True, right_index=True)\n",
    "yermanos_adata = yermanos_adata[yermanos_merged.cell_id]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tcr_cdr3s_aa</th>\n",
       "      <th>tcr_cdr3s_nt</th>\n",
       "      <th>sample</th>\n",
       "      <th>cell_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGAAGGACA-1</th>\n",
       "      <td>TRA:CAVSAGSNYNVLYF;TRB:CASSFWGDSSYEQYF;TRB:CAS...</td>\n",
       "      <td>TRA:TGCGCAGTCAGTGCGGGGTCTAATTACAACGTGCTTTACTTC...</td>\n",
       "      <td>sample3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCACAGCGTC-1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sample3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCACCGAATT-1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sample3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGGTAGCGCTC-1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sample3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGTCCAGGGCT-1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sample3</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTTCATCACCC-2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sample4</td>\n",
       "      <td>4498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAAGTACATGA-2</th>\n",
       "      <td>TRA:CAAIYPGYQNFYF;TRB:CASSDNGNTEVFF</td>\n",
       "      <td>TRA:TGTGCAGCAATCTATCCGGGTTACCAGAACTTCTATTTT;TR...</td>\n",
       "      <td>sample4</td>\n",
       "      <td>4499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCACACCATCCT-2</th>\n",
       "      <td>TRA:CAAIYPGYQNFYF;TRB:CASSDHTNTEVFF</td>\n",
       "      <td>TRA:TGTGCAGCAATCTATCCGGGTTACCAGAACTTCTATTTT;TR...</td>\n",
       "      <td>sample4</td>\n",
       "      <td>4500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTGCAGACA-2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sample4</td>\n",
       "      <td>4501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTGTTGGGA-2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sample4</td>\n",
       "      <td>4502</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4503 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                         tcr_cdr3s_aa  \\\n",
       "AAACCTGAGAAGGACA-1  TRA:CAVSAGSNYNVLYF;TRB:CASSFWGDSSYEQYF;TRB:CAS...   \n",
       "AAACCTGCACAGCGTC-1                                                NaN   \n",
       "AAACCTGCACCGAATT-1                                                NaN   \n",
       "AAACCTGGTAGCGCTC-1                                                NaN   \n",
       "AAACCTGTCCAGGGCT-1                                                NaN   \n",
       "...                                                               ...   \n",
       "TTTGGTTTCATCACCC-2                                                NaN   \n",
       "TTTGTCAAGTACATGA-2                TRA:CAAIYPGYQNFYF;TRB:CASSDNGNTEVFF   \n",
       "TTTGTCACACCATCCT-2                TRA:CAAIYPGYQNFYF;TRB:CASSDHTNTEVFF   \n",
       "TTTGTCAGTGCAGACA-2                                                NaN   \n",
       "TTTGTCAGTGTTGGGA-2                                                NaN   \n",
       "\n",
       "                                                         tcr_cdr3s_nt  \\\n",
       "AAACCTGAGAAGGACA-1  TRA:TGCGCAGTCAGTGCGGGGTCTAATTACAACGTGCTTTACTTC...   \n",
       "AAACCTGCACAGCGTC-1                                                NaN   \n",
       "AAACCTGCACCGAATT-1                                                NaN   \n",
       "AAACCTGGTAGCGCTC-1                                                NaN   \n",
       "AAACCTGTCCAGGGCT-1                                                NaN   \n",
       "...                                                               ...   \n",
       "TTTGGTTTCATCACCC-2                                                NaN   \n",
       "TTTGTCAAGTACATGA-2  TRA:TGTGCAGCAATCTATCCGGGTTACCAGAACTTCTATTTT;TR...   \n",
       "TTTGTCACACCATCCT-2  TRA:TGTGCAGCAATCTATCCGGGTTACCAGAACTTCTATTTT;TR...   \n",
       "TTTGTCAGTGCAGACA-2                                                NaN   \n",
       "TTTGTCAGTGTTGGGA-2                                                NaN   \n",
       "\n",
       "                     sample  cell_id  \n",
       "AAACCTGAGAAGGACA-1  sample3        0  \n",
       "AAACCTGCACAGCGTC-1  sample3        1  \n",
       "AAACCTGCACCGAATT-1  sample3        2  \n",
       "AAACCTGGTAGCGCTC-1  sample3        3  \n",
       "AAACCTGTCCAGGGCT-1  sample3        4  \n",
       "...                     ...      ...  \n",
       "TTTGGTTTCATCACCC-2  sample4     4498  \n",
       "TTTGTCAAGTACATGA-2  sample4     4499  \n",
       "TTTGTCACACCATCCT-2  sample4     4500  \n",
       "TTTGTCAGTGCAGACA-2  sample4     4501  \n",
       "TTTGTCAGTGTTGGGA-2  sample4     4502  \n",
       "\n",
       "[4503 rows x 4 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "yermanos_adata.obs = yermanos_merged\n",
    "yermanos_adata.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read Wherry Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 3856 × 31053\n",
       "    var: 'gene_ids', 'feature_types'"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wherry_adata = sc.read_10x_mtx(f'../data/Wherry2022/10X', gex_only=False)\n",
    "wherry_adata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merge Datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2986723/4064421031.py:1: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  satpathy_adata.obs['dataset'] = 'Satpathy'\n",
      "/Genomics/argo/users/ta6403/.conda/envs/ta6403/lib/python3.11/site-packages/anndata/_core/anndata.py:1763: FutureWarning: The AnnData.concatenate method is deprecated in favour of the anndata.concat function. Please use anndata.concat instead.\n",
      "\n",
      "See the tutorial for concat at: https://anndata.readthedocs.io/en/latest/concatenation.html\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "satpathy_adata.obs['dataset'] = 'Satpathy'\n",
    "cui_adata.obs['dataset'] = 'Cui'\n",
    "cui_adata.obs['ident'] = 'Cui 2023'\n",
    "yermanos_adata.obs['dataset'] = 'Yermanos'\n",
    "yermanos_adata.obs['ident'] = 'Yermanos 2022'\n",
    "wherry_adata.obs['dataset'] = 'Wherry'\n",
    "wherry_adata.obs['ident'] = 'Wherry 2022'\n",
    "\n",
    "satpathy_adata.obs['batch'] = 'Sathpathy_'+satpathy_adata.obs['experiment']\n",
    "cui_adata.obs['batch'] = 'Cui_'+cui_adata.obs['batch'].astype(str)\n",
    "yermanos_adata.obs['batch'] = 'Yermanos_'+yermanos_adata.obs['sample'].astype(str)\n",
    "wherry_adata.obs['batch'] = 'Wherry'\n",
    "\n",
    "try:\n",
    "    del satpathy_adata.obs['tissue']\n",
    "    del satpathy_adata.obs['lcmv']\n",
    "    del satpathy_adata.obs['experiment']\n",
    "    del satpathy_adata.obs['UMAP_1']\n",
    "    del satpathy_adata.obs['UMAP_2']\n",
    "    del satpathy_adata.obs['gated_D8_prog']\n",
    "    del satpathy_adata.obs['is_D8_prog']\n",
    "except:\n",
    "    pass\n",
    "\n",
    "adata = satpathy_adata.concatenate(cui_adata, yermanos_adata, wherry_adata, batch_key=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig.ident</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>tcr_cdr3s_aa</th>\n",
       "      <th>tcr_cdr3s_nt</th>\n",
       "      <th>sort</th>\n",
       "      <th>ident</th>\n",
       "      <th>batch</th>\n",
       "      <th>cell_id</th>\n",
       "      <th>dataset</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGAGATCGGGT-1-0</th>\n",
       "      <td>Spleen_Chronic_D21_TetPos</td>\n",
       "      <td>12513.0</td>\n",
       "      <td>3674.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TetPos</td>\n",
       "      <td>TEx_prog</td>\n",
       "      <td>Sathpathy_exp1</td>\n",
       "      <td>0</td>\n",
       "      <td>Satpathy</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGAGCCACTAT-1-0</th>\n",
       "      <td>Spleen_Chronic_D21_TetPos</td>\n",
       "      <td>3212.0</td>\n",
       "      <td>1535.0</td>\n",
       "      <td>TRA:CAAFANSGTYQRF;TRB:CASSEDWVEQYF</td>\n",
       "      <td>TRA:TGTGCAGCTTTTGCAAATTCTGGGACTTACCAGAGGTTT;TR...</td>\n",
       "      <td>TetPos</td>\n",
       "      <td>TEx_KLR</td>\n",
       "      <td>Sathpathy_exp1</td>\n",
       "      <td>1</td>\n",
       "      <td>Satpathy</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGAGCGTTGCC-1-0</th>\n",
       "      <td>Spleen_Chronic_D21_TetPos</td>\n",
       "      <td>4726.0</td>\n",
       "      <td>1959.0</td>\n",
       "      <td>TRA:CAFEVVGQLTF;TRB:CASSFRDSSYEQYF</td>\n",
       "      <td>TRA:TGTGCTTTCGAGGTTGTGGGGCAGCTCACTTTC;TRB:TGTG...</td>\n",
       "      <td>TetPos</td>\n",
       "      <td>TEx</td>\n",
       "      <td>Sathpathy_exp1</td>\n",
       "      <td>2</td>\n",
       "      <td>Satpathy</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGAGGCTCTTA-1-0</th>\n",
       "      <td>Spleen_Chronic_D21_TetPos</td>\n",
       "      <td>3969.0</td>\n",
       "      <td>1722.0</td>\n",
       "      <td>TRA:CALSEPNYNVLYF;TRB:CASSDWDNTEVFF</td>\n",
       "      <td>TRA:TGTGCTCTGAGTGAGCCTAATTACAACGTGCTTTACTTC;TR...</td>\n",
       "      <td>TetPos</td>\n",
       "      <td>TEx</td>\n",
       "      <td>Sathpathy_exp1</td>\n",
       "      <td>3</td>\n",
       "      <td>Satpathy</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGCAAAGCAAT-1-0</th>\n",
       "      <td>Spleen_Chronic_D21_TetPos</td>\n",
       "      <td>4089.0</td>\n",
       "      <td>1833.0</td>\n",
       "      <td>TRA:CAVSIPGSWQLIF;TRB:CASSSRAGNTEVFF</td>\n",
       "      <td>TRA:TGTGCTGTGAGCATCCCTGGCAGCTGGCAACTCATCTTT;TR...</td>\n",
       "      <td>TetPos</td>\n",
       "      <td>TEx_int</td>\n",
       "      <td>Sathpathy_exp1</td>\n",
       "      <td>4</td>\n",
       "      <td>Satpathy</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGAGGTCTGGA-1-3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Wherry 2022</td>\n",
       "      <td>Wherry</td>\n",
       "      <td>85804</td>\n",
       "      <td>Wherry</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGAGTGAACAT-1-3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Wherry 2022</td>\n",
       "      <td>Wherry</td>\n",
       "      <td>85805</td>\n",
       "      <td>Wherry</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGAGTGGCCTC-1-3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Wherry 2022</td>\n",
       "      <td>Wherry</td>\n",
       "      <td>85806</td>\n",
       "      <td>Wherry</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGCAAGAGCTG-1-3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Wherry 2022</td>\n",
       "      <td>Wherry</td>\n",
       "      <td>85807</td>\n",
       "      <td>Wherry</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGTCTTACGTT-1-3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Wherry 2022</td>\n",
       "      <td>Wherry</td>\n",
       "      <td>85808</td>\n",
       "      <td>Wherry</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>85809 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                   orig.ident  \\\n",
       "Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGAGATCG...  Spleen_Chronic_D21_TetPos   \n",
       "Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGAGCCAC...  Spleen_Chronic_D21_TetPos   \n",
       "Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGAGCGTT...  Spleen_Chronic_D21_TetPos   \n",
       "Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGAGGCTC...  Spleen_Chronic_D21_TetPos   \n",
       "Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGCAAAGC...  Spleen_Chronic_D21_TetPos   \n",
       "...                                                                       ...   \n",
       "TTTGTTGAGGTCTGGA-1-3                                                      NaN   \n",
       "TTTGTTGAGTGAACAT-1-3                                                      NaN   \n",
       "TTTGTTGAGTGGCCTC-1-3                                                      NaN   \n",
       "TTTGTTGCAAGAGCTG-1-3                                                      NaN   \n",
       "TTTGTTGTCTTACGTT-1-3                                                      NaN   \n",
       "\n",
       "                                                    nCount_RNA  nFeature_RNA  \\\n",
       "Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGAGATCG...     12513.0        3674.0   \n",
       "Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGAGCCAC...      3212.0        1535.0   \n",
       "Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGAGCGTT...      4726.0        1959.0   \n",
       "Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGAGGCTC...      3969.0        1722.0   \n",
       "Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGCAAAGC...      4089.0        1833.0   \n",
       "...                                                        ...           ...   \n",
       "TTTGTTGAGGTCTGGA-1-3                                       NaN           NaN   \n",
       "TTTGTTGAGTGAACAT-1-3                                       NaN           NaN   \n",
       "TTTGTTGAGTGGCCTC-1-3                                       NaN           NaN   \n",
       "TTTGTTGCAAGAGCTG-1-3                                       NaN           NaN   \n",
       "TTTGTTGTCTTACGTT-1-3                                       NaN           NaN   \n",
       "\n",
       "                                                                            tcr_cdr3s_aa  \\\n",
       "Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGAGATCG...                                   NaN   \n",
       "Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGAGCCAC...    TRA:CAAFANSGTYQRF;TRB:CASSEDWVEQYF   \n",
       "Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGAGCGTT...    TRA:CAFEVVGQLTF;TRB:CASSFRDSSYEQYF   \n",
       "Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGAGGCTC...   TRA:CALSEPNYNVLYF;TRB:CASSDWDNTEVFF   \n",
       "Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGCAAAGC...  TRA:CAVSIPGSWQLIF;TRB:CASSSRAGNTEVFF   \n",
       "...                                                                                  ...   \n",
       "TTTGTTGAGGTCTGGA-1-3                                                                 NaN   \n",
       "TTTGTTGAGTGAACAT-1-3                                                                 NaN   \n",
       "TTTGTTGAGTGGCCTC-1-3                                                                 NaN   \n",
       "TTTGTTGCAAGAGCTG-1-3                                                                 NaN   \n",
       "TTTGTTGTCTTACGTT-1-3                                                                 NaN   \n",
       "\n",
       "                                                                                         tcr_cdr3s_nt  \\\n",
       "Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGAGATCG...                                                NaN   \n",
       "Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGAGCCAC...  TRA:TGTGCAGCTTTTGCAAATTCTGGGACTTACCAGAGGTTT;TR...   \n",
       "Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGAGCGTT...  TRA:TGTGCTTTCGAGGTTGTGGGGCAGCTCACTTTC;TRB:TGTG...   \n",
       "Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGAGGCTC...  TRA:TGTGCTCTGAGTGAGCCTAATTACAACGTGCTTTACTTC;TR...   \n",
       "Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGCAAAGC...  TRA:TGTGCTGTGAGCATCCCTGGCAGCTGGCAACTCATCTTT;TR...   \n",
       "...                                                                                               ...   \n",
       "TTTGTTGAGGTCTGGA-1-3                                                                              NaN   \n",
       "TTTGTTGAGTGAACAT-1-3                                                                              NaN   \n",
       "TTTGTTGAGTGGCCTC-1-3                                                                              NaN   \n",
       "TTTGTTGCAAGAGCTG-1-3                                                                              NaN   \n",
       "TTTGTTGTCTTACGTT-1-3                                                                              NaN   \n",
       "\n",
       "                                                      sort        ident  \\\n",
       "Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGAGATCG...  TetPos     TEx_prog   \n",
       "Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGAGCCAC...  TetPos      TEx_KLR   \n",
       "Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGAGCGTT...  TetPos          TEx   \n",
       "Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGAGGCTC...  TetPos          TEx   \n",
       "Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGCAAAGC...  TetPos      TEx_int   \n",
       "...                                                    ...          ...   \n",
       "TTTGTTGAGGTCTGGA-1-3                                   NaN  Wherry 2022   \n",
       "TTTGTTGAGTGAACAT-1-3                                   NaN  Wherry 2022   \n",
       "TTTGTTGAGTGGCCTC-1-3                                   NaN  Wherry 2022   \n",
       "TTTGTTGCAAGAGCTG-1-3                                   NaN  Wherry 2022   \n",
       "TTTGTTGTCTTACGTT-1-3                                   NaN  Wherry 2022   \n",
       "\n",
       "                                                             batch  cell_id  \\\n",
       "Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGAGATCG...  Sathpathy_exp1        0   \n",
       "Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGAGCCAC...  Sathpathy_exp1        1   \n",
       "Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGAGCGTT...  Sathpathy_exp1        2   \n",
       "Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGAGGCTC...  Sathpathy_exp1        3   \n",
       "Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGCAAAGC...  Sathpathy_exp1        4   \n",
       "...                                                            ...      ...   \n",
       "TTTGTTGAGGTCTGGA-1-3                                        Wherry    85804   \n",
       "TTTGTTGAGTGAACAT-1-3                                        Wherry    85805   \n",
       "TTTGTTGAGTGGCCTC-1-3                                        Wherry    85806   \n",
       "TTTGTTGCAAGAGCTG-1-3                                        Wherry    85807   \n",
       "TTTGTTGTCTTACGTT-1-3                                        Wherry    85808   \n",
       "\n",
       "                                                     dataset sample  \n",
       "Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGAGATCG...  Satpathy    NaN  \n",
       "Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGAGCCAC...  Satpathy    NaN  \n",
       "Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGAGCGTT...  Satpathy    NaN  \n",
       "Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGAGGCTC...  Satpathy    NaN  \n",
       "Spleen_Chronic_D21_CD8_TetPos_gex_AAACCTGCAAAGC...  Satpathy    NaN  \n",
       "...                                                      ...    ...  \n",
       "TTTGTTGAGGTCTGGA-1-3                                  Wherry    NaN  \n",
       "TTTGTTGAGTGAACAT-1-3                                  Wherry    NaN  \n",
       "TTTGTTGAGTGGCCTC-1-3                                  Wherry    NaN  \n",
       "TTTGTTGCAAGAGCTG-1-3                                  Wherry    NaN  \n",
       "TTTGTTGTCTTACGTT-1-3                                  Wherry    NaN  \n",
       "\n",
       "[85809 rows x 11 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create unique cell id\n",
    "adata.obs['cell_id'] = range(len(adata.obs))\n",
    "\n",
    "adata.obs.loc[~adata.obs['tcr_cdr3s_aa'].str.contains('TRA:', na=False), 'tcr_cdr3s_aa'] = np.nan\n",
    "adata.obs.loc[~adata.obs['tcr_cdr3s_aa'].str.contains('TRA:', na=False), 'tcr_cdr3s_nt'] = np.nan\n",
    "adata.obs.loc[~adata.obs['tcr_cdr3s_aa'].str.contains('TRB:', na=False), 'tcr_cdr3s_aa'] = np.nan\n",
    "adata.obs.loc[~adata.obs['tcr_cdr3s_aa'].str.contains('TRB:', na=False), 'tcr_cdr3s_nt'] = np.nan\n",
    "\n",
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tcr_cdr3s_aa\n",
       "TRA:CAFEVVGQLTF;TRB:CASSFRDSSYEQYF                            3748\n",
       "TRA:CAVSAWVGDNSKLIW;TRB:CASSRDSLTEVFF                         1839\n",
       "TRA:CAQLEGADRLTF;TRB:CASSDHTNTEVFF                            1300\n",
       "TRA:CAMSDLSNNAGAKLTF;TRB:CASSRDNNYAEQFF                       1007\n",
       "TRA:CAALNNYAQGLTF;TRB:CASSLIRTGGYEQYF                          959\n",
       "                                                              ... \n",
       "TRA:CAVSLDRGSALGRLHF;TRA:CALGPDTNAYKVIF;TRB:CASSIGTSYAEQFF       1\n",
       "TRA:CAANYNQGKLIF;TRA:CAMREGEQGTGSKLSF;TRB:CASSFDWGGGQDTQYF       1\n",
       "TRA:CAANTGYQNFYF;TRB:CASSQDREDSDYTF                              1\n",
       "TRA:CALSERYNQGKLIF;TRB:CASSLNLEVFF                               1\n",
       "TRA:CAASSNSNNRIFF;TRB:CASSDHDNTEVFF                              1\n",
       "Name: count, Length: 10757, dtype: int64"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.tcr_cdr3s_aa.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write('../data/datasets_combined.h5ad')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:scvi]",
   "language": "python",
   "name": "conda-env-scvi-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
